sound/soc/sof/intel/hda-dsp.c

```diff
@@ -446,26 +446,38 @@ static int hda_dsp_set_D0_state(struct snd_sof_dev *sdev,
 		    !hda_enable_trace_D0I3_S0 ||
 		    sdev->system_suspend_target != SOF_SUSPEND_NONE)
 			flags = HDA_PM_NO_DMA_TRACE;
+
+		ret = hda_dsp_send_pm_gate_ipc(sdev, flags);
+		if (ret < 0) {
+			dev_err(sdev->dev,
+				"error: PM_GATE ipc error %d\n", ret);
+			return -EIO;
+		}
+
+		/* update D0I3C register */
+		ret = hda_dsp_update_d0i3c_register(sdev, value);
+		if (ret < 0)
+			return ret;
 	} else {
 		/* prevent power gating in D0I0 */
 		flags = HDA_PM_PPG;
-	}
 
-	/* update D0I3C register */
-	ret = hda_dsp_update_d0i3c_register(sdev, value);
-	if (ret < 0)
-		return ret;
```

Comment:
Why do we need different sequence for d0i3 vs !d0i3?
This patch basically does: if d0i3 then ipc_first else ipc_last:
d0i3:  ipc, register
!d0i3: register, ipc

How did it worked in the past when it was always register, ipc sequence?

Reply:
@ujfalusi i think it makes sense to reverse the order for entry/exit. As for how it workd, my only guess is that it probably doesnt matter